arc/alkali_atom_functions.py

```diff
@@ -787,7 +787,20 @@ def getEnergy(self, n, l, j, s=0.5):
 
         # else, use quantum defects
         defect = self.getQuantumDefect(n, l, j, s=s)
-        return -self.scaledRydbergConstant / ((n - defect) ** 2)
+
+        
+
+        if l <= 3:
+            return -self.scaledRydbergConstant / ((n - defect) ** 2)
+        elif l == 4:
+            FS = (2*self.scaledRydbergConstant)*pow(scipy.constants.alpha,2)/(2*l*(l+1)*pow(n,3))
+            LI = -FS/(l+0.5)
+            return -self.scaledRydbergConstant / ((n - defect) ** 2) -0.5*LI*(j*(j+1)-l*(l+1)-s*(s+1))
+        else:
+            FS = (2*self.scaledRydbergConstant)*pow(scipy.constants.alpha,2)/(2*l*(l+1)*pow(n,3))
+            LI = -FS/(l+0.5)
+            return -self.scaledRydbergConstant / ((n) ** 2) -0.5*LI*(j*(j+1)-l*(l+1)-s*(s+1)) - 2*self.scaledRydbergConstant*defect
```

Comment:
What is difference between `l==4` and `l>=4` cases here? It seems to me that second one is just expansion for small `defect` value right? If so, I would rather keep one case (without expansion). Please comment.

Reply:
What is meaning of `FS` and `LS` ? Note that we try in general to stick to convention of `camelCase` naming of variables.

Additionally, where can you please add in code comment that references this equations in the literature? I am maybe looking at wrong place, but I cannot find them.

Reply:
> What is difference between `l==4` and `l>=4` cases here? It seems to me that second one is just expansion for small `defect` value right? If so, I would rather keep one case (without expansion). Please comment.

I think that the expansion is needed in the case of l>=5 (not l>=4) as we are assuming that the energy difference is due to a perturbation and that there are no other effects (penetration, exchange interaction). i.e. We are not really calculating a quantum defect but an energy correction. I have simply used `getQuantumDefect'` as a way of obtaining a change in energy using the stored n,l,j values.

If you would like, it may be more appropriate to define one (or two) new functions to remove this confusion. i.e. `getPolarisationEnergy' (and `getHydrogenFS' splitting for the other) so that it is clear l>5 is treated differently. 

That is, a correction to the hydrogenic energy level from some polarisation energy calculated from perturbation theory.

Reply:
> What is meaning of `FS` and `LS` ? Note that we try in general to stick to convention of `camelCase` naming of variables.
> 
> Additionally, where can you please add in code comment that references this equations in the literature? I am maybe looking at wrong place, but I cannot find them.

Apologies, FS and LS are constants that I defined that are used in the convention of `The Theory Of Atomic Spectra
by [Shortley, G.h] (pg. 193 onwards).` and from https://staff.fnwi.uva.nl/j.t.m.walraven/walraven/Publications_files/2023-AtomicPhysics.pdf section 4.5.2 onwards. FS and LS were mainly defined to improve the readability of the code (but are completely arbitrary). They could be renamed to fineStructSplit (for FS).

This is a bit hard to follow so I propose the following changes (following Foot. Atomic Physics pg. 37-38. So the code now reads.

`if l <= 3:
            return -self.scaledRydbergConstant / ((n - defect) ** 2)
        elif l == 4:
            fineStructSplit = -(2*self.scaledRydbergConstant)*pow(scipy.constants.alpha,2)/(2*l*(l+0.5)*(l+1)*pow(n,3))

            if j == l + 0.5:
                 
                 fineStructSplit *= 0.5*l

            elif j == l - 0.5:

                fineStructSplit *= -1*0.5*(l+1)

            return -self.scaledRydbergConstant / ((n - defect) ** 2) - fineStructSplit
        else:
            fineStructSplit = -(2*self.scaledRydbergConstant)*pow(scipy.constants.alpha,2)/(2*l*(l+0.5)*(l+1)*pow(n,3))

            if j == l + 0.5:
                 
                 fineStructSplit *= 0.5*l

            elif j == l - 0.5:

                fineStructSplit *= -1*0.5*(l+1)
                
            return -self.scaledRydbergConstant / ((n - defect) ** 2) - fineStructSplit - 2*self.scaledRydbergConstant*defect`

Reply:
> If you would like, it may be more appropriate to define one (or two) new functions to remove this confusion. i.e. `getPolarisationEnergy` (and `getHydrogenFS` splitting for the other) so that it is clear l>5 is treated differently.

@johnny-sa  Sorry for coming late back to this. Introducing these two methods would be good, and then changes should be to `getEnergy` method, not `getQuantumDefect` because quantum defects are, as you said, unchanged, and people might be using them for other purposes then calculating energy levels. Do you have time to do it? If not, I can try refactoring your code, and then you can review it.

Reply:
@johnny-sa let me know what works best for you. I have finally some extra time to work on this if needed.

Reply:
> > If you would like, it may be more appropriate to define one (or two) new functions to remove this confusion. i.e. `getPolarisationEnergy` (and `getHydrogenFS` splitting for the other) so that it is clear l>5 is treated differently.
> 
> @johnny-sa Sorry for coming late back to this. Introducing these two methods would be good, and then changes should be to `getEnergy` method, not `getQuantumDefect` because quantum defects are, as you said, unchanged, and people might be using them for other purposes then calculating energy levels. Do you have time to do it? If not, I can try refactoring your code, and then you can review it.

Hi apologies for the late reply, I can look at doing this presently unless you have already. It's been a while since I've thought about this and have had some time to think it over. My previous comments were a bit all over the place.

I think the best option would be to keep calculating the polarisation 'correction' in the quantum defect. You can still represent the polarisation energy as a quantum defect [1]. This way, if someone wanted to use `getQuantumDefect` of high L states, it would still return a number. 

Then I think it would best to add the hydrogenic (fine-structure) splitting and relativistic correction for states with l>5 to the `getEnergy`, as there is no core penetration with these states. I can define a new function with 'getHydrogenicCorrection` to incorporate these. (these are MHz level shifts and would not change the value of $\delta_{\ell}$ that much). 

[1] - Freeman, Richard R., and Daniel Kleppner. Physical Review A 14.5 (1976): 1614.

Reply:
Hi @nikolasibalic, I added these changes and put two citations for both. 
https://journals.aps.org/pra/abstract/10.1103/PhysRevA.14.1614 for the general method of using polarisability for high l states. And
https://journals.aps.org/pra/abstract/10.1103/PhysRevA.100.012501 for using relativistic and FS corrections for these high l states.